### ISSUE TITLE
Create a Proprerty object that can be used instead of a BehaviorSubject

### ISSUE THREAD

**Issue by mikebaum** [mikebaum/RxUI#1]
As has been described in the project description, the core of this library is based around the concept of a property. This issue tracks the work to create that abstraction. Initially a property has the following api:

```
interface Property {
    void setValue(M value);
    void accept(M value);
    M get();
    void dispose();
    boolean hasObservers();
    Subscription bind(PropertyObservable<M> propertyToBindTo);
    Subscription synchronize(Property<M> propertyToSynchornizeWith);
    Subscription observe(Consumer<M> onChanged, Runnable onDisposed);
    Subscription onChanged(Consumer<M> onChanged);
    Subscription onDisposed(Runnable onDisposedAction);
    Observable<M> asObservable();
    static Property<M> create(M initialValue);
    static Property<M> create(PropertySource<M> propertySource);
    static Property<Optional<M>> createOptional();
    static Property<Optional<M>> createOptional(M initialValue);
    static Property<T> fromSubject(BehaviorSubject<T> subject);
}
```

A property uses the RxJava `Subscription` object to provide a means to terminate a subscription. This was done for easier interop with RxJava, but could change.

Missing from the initial commit are the following methods:

```
    void reset();
    Observable isDirty();
```

The `isDirty` method could return an Observable, but this can be discussed. The alternative would be to create our own Observable, but that might be overkill. In fact I will be committing a `PropertyObservable` interface, but I am on the fence about it and it needs to be discussed.

The pull request adds a `PropertySource` hierarchy that in my opinion right now could be factored differently, but it's at a start.

The `PropertyDispatcher` is also something that requires consideration, as changes to the `PropertySource` abstraction would no doubt mean updating the dispatcher.

**Comment by mikebaum** [comment]
decided to add the reset method, I rebased the pull request.

**Comment by mikebaum** [comment]
An implementation of this Issue has been included in Issue #3. Closing this Issue.
